### ISSUE TITLE
[Bug]: PaDiM models trained with custom dataset are not deterministic

### ISSUE THREAD

**Issue by retrok94** [openvinotoolkit/anomalib#1735]
### Describe the bug

Training PaDiM models (and possibly other models as well, I have not tried those) with the same parameters and seeds results in different results during inference.
The same PaDiM model is deterministic when I run inference on it multiple times. The problem only occurs if I train a new model and run inference again.

I have tried the various settings and trained both on CPU and GPU, but the results remain non-deterministic.

Things I have tried (besides setting the seed for the Folder datamodule):
- Trained on GPU
- Trained on CPU
- Set num_workers to 0
- Passed {'deterministic': True} as **kwargs for Trainer to Engine API call
- Set the following parameters/seeds to prevent non-determinism (some for GPU and some for CPU):
> os.environ["CUBLAS_WORKSPACE_CONFIG"] = ":4096:8"
> os.environ["MKL_NUM_THREADS"] = "1"
> os.environ["NUMEXPR_NUM_THREADS"] = "1"
> os.environ["OMP_NUM_THREADS"] = "1"
> os.environ["OPENBLAS_NUM_THREADS"] = "1"
> os.environ["PL_GLOBAL_SEED"] = str(seed)
> os.environ["PL_SEED_WORKERS"] = f"{int(True)}"
> random.seed(seed)
> np.random.seed(seed)
> torch.manual_seed(seed)
> torch.cuda.manual_seed_all(seed)
> torch.use_deterministic_algorithms(mode=True)
> torch.backends.cudnn.deterministic = True
> torch.backends.cudnn.benchmark = False

Is this impossible to prevent? Or is there another setting I am missing?

### Dataset

Folder

### Model

PADiM

### Steps to reproduce the behavior

- Install anomalib 1.0.0dev0
- Run script with suggested API calls and Folder datamodule that gets passed my seed (42). Engine is setup with task=TaskType.CLASSIFICATION and image_metrics=['F1AdaptiveThreshold']
- Export model in torch format
- Run inference with TorchInferencer and same settings for Folder datamodule.

### OS information

OS information:
- OS: Windows 11
- Python version: 3.10.0
- Anomalib version: 1.0.0dev
- PyTorch version: 2.1.2
- CUDA/cuDNN version: 12.1
- GPU models and configuration: RTX 4090
- Custom Dataset

### Expected behavior

Identical results during inference for every PaDiM model trained with same parameters and seed.

### Pip/GitHub

GitHub

### Configuration YAML

```yaml
(API calls)
```

### Logs

```shell
-
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by alexriedel1** [comment]
Have you tried to hard determine the test split images? E.g. setting test_split_ratio = 1.0 and val_split_ratio = 1.0 (These might not be useful settings and I'm not sure if it will even work, but I faced some issues that the test and val split were not deterministic even though everything was set to deterministic before. Which might result from the fact the manual_seed is called again here: https://github.com/openvinotoolkit/anomalib/blob/13b6a7cb3cb1f7dc0172923779602c4ac380ab0c/src/anomalib/data/utils/split.py#L117-L122

**Comment by alexriedel1** [comment]
Also you could try to debug non deterministic actions https://pytorch.org/docs/stable/generated/torch.set_deterministic_debug_mode.html#torch.set_deterministic_debug_mode

use `CUBLAS_WORKSPACE_CONFIG=:16:8` as well as `CUBLAS_WORKSPACE_CONFIG=:4096:8`

**Comment by retrok94** [comment]
Update:
I managed to get everything deterministic by setting the above mentioned parameters/seeds again in the torch_model.py file of the padim model. Not sure if some parameters have to be set there to have any effect (currently I'm setting them in my own script with the API calls to anomalib) or if some of them are reset/changed at some point between me calling anomlib and anomalib calling the torch model.
I'll have to investigate which settings exactly need to be set to speed up the training again, since it is (expectedly) very slow this way. But I'll close this issue for now since the main problem is solved.
